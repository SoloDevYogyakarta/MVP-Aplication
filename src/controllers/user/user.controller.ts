import { Request, Response } from 'express';
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpStatus,
  Post,
  Query,
  Req,
  Res,
  UploadedFile,
  UseGuards,
  UseInterceptors,
} from '@nestjs/common';
import { UserRepository } from '../../repository/user/user.repository';
import {
  LoginField,
  RegisterField,
} from '../../validators/user/user.validator';
import { UserService } from '../../services/user/user.service';
import { AuthGuard } from '../../middleware/guards.middleware';
import { createpath } from '../../utils/system/system';
import { omit } from 'lodash';
import { DyanmicQuery } from '../../validators/query/product.query';
import { FileInterceptor } from '@nestjs/platform-express';
import { diskStorage } from '../../utils/multer/multer';
import { CustomRequest } from '../../types/custom-request.type';

@Controller('user')
export class UserController {
  constructor(
    private readonly repository: UserRepository,
    private readonly service: UserService,
  ) {}

  @Post('login')
  async login(
    @Req() req: Request,
    @Res() res: Response,
    @Body() body: LoginField,
  ) {
    const result = await this.service.login(body);
    return res.status(result.status).json(result);
  }

  @Post()
  async create(
    @Req() req: Request,
    @Res() res: Response,
    @Body() body: RegisterField,
  ) {
    const result = await this.service.create(body);
    createpath('../../database/dataTxt/user-http-entity.txt', result.result);
    return res.status(result.status).json(omit(result, ['result']));
  }

  @Post('reset/password')
  reset(@Req() req: Request, @Res() res: Response) {
    return res.status(HttpStatus.OK).json('reset');
  }

  @UseInterceptors(FileInterceptor('file', { storage: diskStorage }))
  @UseGuards(AuthGuard)
  @Post(':id')
  async update(
    @Req() req: Request,
    @Res() res: Response,
    @UploadedFile() file: Express.Multer.File,
  ) {
    const result = await this.service.update(req.params.id, req.body, file);
    return res.status(result.status).json(result);
  }

  @Delete(':id')
  async destroy(@Req() req: Request, @Res() res: Response) {
    const result = await this.service.destroy(req.params.id);
    return res.status(result.status).json(result);
  }

  @UseGuards(AuthGuard)
  @Get()
  async list(
    @Query() query: DyanmicQuery,
    @Req() req: Request,
    @Res() res: Response,
  ) {
    const result = await this.repository.findAll(query);
    return res.status(HttpStatus.OK).json(result);
  }

  @UseGuards(AuthGuard)
  @Get(':id')
  async detail(@Req() req: Request, @Res() res: Response) {
    const result = await this.repository.findOne(req.params.id);
    return res.status(HttpStatus.OK).json(result);
  }

  @UseGuards(AuthGuard)
  @Get('access/me')
  async me(@Req() req: CustomRequest, @Res() res: Response) {
    return res.status(HttpStatus.OK).json(req.user.data);
  }
}
