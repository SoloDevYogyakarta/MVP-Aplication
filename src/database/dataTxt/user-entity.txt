{"id":244,"username":"Hattie_Waters23","password":"$2b$10$KkrxxQz79dBpcJHsOryBjeta/ltDU0p3vo2cTeYrBcBxpblrlmh86","updatedAt":"2023-10-21T09:54:52.365Z","createdAt":"2023-10-21T09:54:52.365Z","public_id":"ehRhduFOwW00p5qF18jzw","email":null,"phone_number":null}