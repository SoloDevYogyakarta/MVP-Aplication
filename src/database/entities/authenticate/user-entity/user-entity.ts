import { nanoid } from 'nanoid';
import { DataTypes, Model, ModelAttributes, Optional } from 'sequelize';
import { SuperTest } from 'supertest';
import { BaseEntity } from '../../base-entity/base-entity';
import { sequelize } from '../../entity';

export class UserEntity extends BaseEntity {
  plat_number: string;
  phone_number: string;
  name: string;
  motor: string;
  year_production: number;
  address: string;
  role: string;
  password: string;
}

export type UserCreationAttribute = Optional<UserEntity, 'id'>;
export interface UserInstance
  extends Model<UserCreationAttribute, UserEntity>,
    UserEntity {}
export const userEntity = sequelize.define<UserInstance>(
  'USER',
  {
    plat_number: DataTypes.STRING,
    phone_number: DataTypes.STRING,
    name: DataTypes.STRING,
    motor: DataTypes.STRING,
    year_production: DataTypes.INTEGER,
    address: DataTypes.STRING,
    role: DataTypes.STRING,
    password: DataTypes.STRING,
  } as ModelAttributes<UserInstance>,
  {
    tableName: 'USER',
    schema: 'AUTHENTICATE',
  },
);
