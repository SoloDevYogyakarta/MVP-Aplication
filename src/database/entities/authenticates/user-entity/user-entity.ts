import { DataTypes, Model, ModelAttributes, Optional } from 'sequelize';
import { nanoid } from 'nanoid';
import bcrypt from 'bcrypt';
import { sequelize } from '../../entity';
import { BaseEntity } from '../../base-entity/base-entity';

export class UserEntity extends BaseEntity {
  username: string;
  email: string;
  phone_number: number;
  password: string;
  role: string;
  file_id: string;
}

export type UserCreationAttribute = Optional<UserEntity, 'id'>;
export interface UserInstance
  extends Model<UserCreationAttribute, UserEntity>,
    UserEntity {}
export const userEntity = sequelize.define<UserInstance>(
  'USER',
  {
    public_id: DataTypes.STRING,
    username: DataTypes.STRING,
    email: DataTypes.STRING,
    phone_number: DataTypes.INTEGER,
    role: DataTypes.STRING,
    password: DataTypes.STRING,
    file_id: DataTypes.STRING,
  } as ModelAttributes<UserInstance>,
  {
    tableName: 'USER',
    schema: 'AUTHENTICATE',
    hooks: {
      beforeCreate(attributes, options) {
        const instance = attributes;
        instance.public_id = nanoid();
        instance.password = bcrypt.hashSync(instance.password, 10);
      },
    },
  },
);
