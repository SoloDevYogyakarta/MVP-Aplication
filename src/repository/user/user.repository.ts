import { Injectable } from '@nestjs/common';
import {
  userEntity,
  UserInstance,
} from '../../database/entities/authenticates/user-entity/user-entity';

@Injectable()
export class UserRepository {
  async findAll(): Promise<UserInstance[]> {
    const result = await userEntity.findAll();
    return result;
  }

  async findOne(public_id: string): Promise<UserInstance> {
    const result = await userEntity.findOne({ where: { public_id } });
    return result;
  }
}
