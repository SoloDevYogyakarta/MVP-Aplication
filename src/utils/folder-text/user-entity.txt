{"id":1057,"plat_number":"B 6929 SHA","phone_number":"540-311-3034","name":"Yvette Tremblay","motor":"Sleek Soft Mouse","address":"39389 Franey Courts","year_production":2861,"role":"admin","password":"$2b$10$Q3pPGa3yL/6IFbR/nUsj8uq2hFGtAbDTUkMqX5A.FyonjCXibU58K","updatedAt":"2023-11-06T15:51:08.666Z","createdAt":"2023-11-06T15:51:08.666Z"}