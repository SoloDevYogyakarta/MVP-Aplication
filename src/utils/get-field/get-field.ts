import { GetFieldType } from 'src/types/get-field.type';
import { readpath } from '../system/system';

export function getField(filename?: GetFieldType) {
  try {
    if (filename === 'fake') {
      return readpath(`../../database/dataTxt/${filename}.txt`);
    }
    return JSON.parse(readpath(`../../database/dataTxt/${filename}.txt`));
  } catch (err) {
    throw new Error(err);
  }
}
