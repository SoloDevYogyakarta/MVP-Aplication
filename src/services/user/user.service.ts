import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { userEntity } from '../../database/entities/authenticates/user-entity/user-entity';
import {
  LoginField,
  RegisterField,
} from '../../validators/user/user.validator';
import bcrypt from 'bcrypt';
import { JwtService } from '@nestjs/jwt';
import { omit } from 'lodash';
import { createpath, removepath } from '../../utils/system/system';
import env from '../../utils/env/env';
import { Op } from 'sequelize';
import { fileEntity } from '../../database/entities/commons/file-entity/file-entity';

@Injectable()
export class UserService {
  constructor(public jwtService: JwtService) {}

  private readonly regex = /^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/g;

  public async login(field: LoginField) {
    let where: { [key: string]: string } = { username: field.token };
    if (this.regex.test(field.token)) {
      where = { email: field.token };
      delete where['username'];
    }
    const findOne = await userEntity.findOne({ where });
    if (!findOne) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: 'Username or password inccorect',
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    const check = await bcrypt.compareSync(field.password, findOne.password);
    if (!check) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: 'Wrong password',
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    const token = this.jwtService.sign(
      { data: findOne },
      { secret: env['SECRET'] },
    );
    return { accessToken: token, status: HttpStatus.OK };
  }

  async create(field: RegisterField) {
    let where = [{ username: '' }, { email: '' }];
    if (field.username) {
      where = [...where, { username: field.username }];
    }
    if (field.email) {
      where = [...where, { email: field.email }];
    }

    const findOne = await userEntity.findOne({
      where: {
        [Op.or]: where,
      },
    });
    if (findOne) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: 'Username already exists, please choose another one',
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    if (field.password !== field.confirmation) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: "Password don't match, please check again",
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    const file = await fileEntity.create({});
    file.save();
    const create = await userEntity.create(
      omit({ ...field, file_id: file.public_id }, ['confirmation']),
    );
    createpath(`../../database/dataTxt/${'user-service-entity.txt'}`, create);
    create.save();
    return {
      result: create,
      status: HttpStatus.CREATED,
      message: 'Account has been created',
    };
  }

  async update(
    public_id: string,
    field: { username: string; password: string },
    file: Express.Multer.File,
  ) {
    const findOne = await userEntity.findOne({ where: { public_id } });
    if (!findOne) {
      throw new HttpException('false', HttpStatus.INTERNAL_SERVER_ERROR);
    }
    const check = await bcrypt.compareSync(field.password, findOne.password);
    if (!check) {
      throw new HttpException(
        { status: HttpStatus.BAD_REQUEST, message: 'Wrong password' },
        HttpStatus.BAD_REQUEST,
      );
    }
    findOne.username = field.username;
    findOne.save();

    const fileEnt = await fileEntity.findOne({
      where: { public_id: findOne.file_id },
    });
    if (file?.originalname) {
      fileEnt.filename = file.filename;
      fileEnt.originalname = file.originalname;
      if (fileEnt.filepath) {
        try {
          removepath(`../..${fileEnt.filepath}`);
        } catch (err) {
          // empty
        }
      }
      fileEnt.filepath = file.path.split('/src')[1];
      fileEnt.type = file.mimetype.split('/')[0];
      fileEnt.save();
    }
    return { status: HttpStatus.OK, message: 'Account has been updated' };
  }

  async destroy(public_id: string) {
    const findOne = await userEntity.findOne({ where: { public_id } });
    if (!findOne) {
      throw new HttpException(
        {
          status: HttpStatus.BAD_REQUEST,
          message: 'Account not found',
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    findOne.destroy();
    return { status: HttpStatus.OK, message: 'Account has been deleted' };
  }
}
