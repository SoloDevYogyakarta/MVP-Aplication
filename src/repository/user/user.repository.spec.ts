import { UserInstance } from '../../database/entities/authenticates/user-entity/user-entity';
import { getField } from '../../utils/get-field/get-field';
import { UserRepository } from './user.repository';

describe('UserRepository', () => {
  let repository: UserRepository;

  beforeEach(() => {
    repository = new UserRepository();
  });

  it('should to be defined', () => expect(repository).toBeDefined());

  it('render correctly', () => expect(repository).toMatchSnapshot());

  it('findOne', async () => {
    const result = getField('user-http-entity') as UserInstance;
    const findOne = await repository.findOne(result.public_id);
    expect(findOne.public_id).toEqual(result.public_id);
  });

  it('findAll', async () => {
    const findAll = await repository.findAll();
    expect(findAll.length).not.toEqual(0);
  });
});
