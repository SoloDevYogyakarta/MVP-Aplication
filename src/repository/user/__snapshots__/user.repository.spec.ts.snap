// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRepository render correctly 1`] = `UserRepository {}`;
